src/irmin-pack/pack_key_intf.ml

```diff
@@ -17,19 +17,60 @@
 open! Import
 
 module type S = sig
-  type hash
-
-  include Irmin.Key.S with type t = hash and type hash := hash
+  include Irmin.Key.S
 
   val null : t
+
+  val unfindable_of_hash : hash -> t
+  (** [unfindable_of_hash h] is a key [k] such that [to_hash k = h], with an
+      unspecified internal representation. This function enables an efficient
+      implmentation of "portable" inodes, but is otherwise unused. Attempting to
+      dereference a key constructed in this way results in undefined behaviour. *)
 end
 
 module type Sigs = sig
-  type 'hash t = 'hash
+  type 'hash t
   (** The type of {i keys} referencing values stored in the [irmin-pack]
       backend. *)
 
-  module type S = S
+  (** The internal state of a key (read with {!inspect}).
+
+      Invariant: keys of the form {!Indexed} always reference values that have
+      entries in the index (as otherwise these keys could not be dereferenced). *)
+  type 'hash state = private
+    | Direct of { hash : 'hash; offset : int63; length : int }
+        (** A "direct" pointer to a value stored at [offset] in the pack-file
+            (with hash [hash] and length [length]). Such keys can be
+            dereferenced from the store with a single IO read, without needing
+            to consult the index.
+
+            They are built in-memory (e.g. after adding a fresh value to the
+            pack file), but have no corresponding encoding format, as the pack
+            format keeps length information with the values themselves.
+
+            When decoding a inode, which references its children as single
+            offsets, we fetch the length information of the child at the same
+            time as fetching its hash (which we must do anyway in order to do an
+            integrity check), creating keys of this form. *)
+    | Indexed of 'hash
+        (** A pointer to an object in the pack file that is indexed. Reading the
+            object necessitates consulting the index, after which the key can be
+            promoted to {!Direct}.
+
+            Such keys result from decoding pointers to other store objects
+            (nodes or commits) from commits or from the branch store. *)
+
+  val inspect : 'hash t -> 'hash state
+  val v_direct : hash:'h -> offset:int63 -> length:int -> 'h t
```

Comment:
Just out of interest, "v" seems to be used as a "create" shorthand throughout the codebase. What does it stand for?

Reply:
I would say: `type t` for type, `val t` for typerepr, `val v` for value

Reply:
I believe the `v` naming scheme for constructors was originally borrowed from Daniel Bünzli's libraries (e.g. [`astring`](https://erratique.ch/software/astring/doc/Astring/String/index.html#val-v)).